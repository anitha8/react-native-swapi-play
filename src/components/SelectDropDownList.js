/**
 * @flow
 */
import React, { Component } from "react";
import {
    Text,
    View,
    TextInput,
    TouchableWithoutFeedback,
    FlatList,
    Image, StyleSheet,
} from "react-native";
import PropTypes from "prop-types";
import find from "lodash/find";
import get from "lodash/get";
import ListItem from "./ListItem";
import usernameImg from '../images/username.png';
import passwordImg from '../images/password.png';

class SelectDropDownList extends Component {
    static propTypes = {
        selectedItems: PropTypes.array,
        items: PropTypes.array.isRequired,
        uniqueKey: PropTypes.string,
        onSelectedItemsChange: PropTypes.func.isRequired,
        placeHolderText: PropTypes.string,
        displayKey: PropTypes.string,
        getSelectedItems: PropTypes.func,
    };

    static defaultProps = {
        selectedItems: [],
        items: [],
        uniqueKey: "id",
        onSelectedItemsChange: () => {
        },
        placeHolderText: "Select",
        displayKey: "name",
    };

    constructor(props) {
        super(props);
        this.state = {
            selector: false,
            searchTerm: "",
            textColor: '#f0f1f1'
        };
    }

    findItem = itemKey => {
        const { items, uniqueKey } = this.props;
        return find(items, singleItem => singleItem[uniqueKey] === itemKey) || {};
    };

    onChangeInput = value => {
        this.setState({ searchTerm: value });
    };

    getSelectLabel = () => {
        const {
            placeHolderText,
            selectedItems,
            displayKey,
        } = this.props;

        const item = selectedItems[0];
        const foundItem = this.findItem(item);
        const getItemname = get(foundItem, displayKey);
        return getItemname || placeHolderText;
    };

    toggleSelector = () => {
        this.setState({
            selector: !this.state.selector,
        });
        const {
            selectedItems,
            displayKey,
            placeHolderText,
        } = this.props;
        const item = selectedItems[0];
        const foundItem = this.findItem(item);
        const getItemname = get(foundItem, displayKey);
        this.setState({ textColor: getItemname != placeHolderText && 'black' || '#f0f1f1' });
    };

    submitSelection = () => {
        this.setState({ searchTerm: "" });
        this.toggleSelector();
    };

    itemSelected = item => {
        const { uniqueKey, selectedItems } = this.props;
        return !!find(selectedItems, singleItem => item[uniqueKey] === singleItem);
    };

    toggleItem = item => {
        const {
            uniqueKey,
            onSelectedItemsChange,
        } = this.props;

        this.submitSelection();
        onSelectedItemsChange([item[uniqueKey]]);
    };

    getRow = item => {
        const { displayKey } = this.props;
        const selectedItemFlag = this.itemSelected(item);

        return (
            <ListItem
                selectedItemFlag={selectedItemFlag}
                onPress={() => this.toggleItem(item)}
                itemText={item[displayKey]}
                gradientColor={item.colorVal}
            />
        );
    };

    filterItems = searchTerm => {
        const { items, displayKey } = this.props;
        const filteredItems = [];
        items.forEach(item => {
            const parts = searchTerm.trim().split(/[ \-:]+/);
            const regex = new RegExp(`(${parts.join("|")})`, "ig");
            if (regex.test(get(item, displayKey))) {
                filteredItems.push(item);
            }
        });
        return filteredItems;
    };

    renderItems = () => {
        const { items, uniqueKey, selectedItems } = this.props;
        const { searchTerm } = this.state;
        let component = null;
        const renderItems = searchTerm
            ? this.filterItems(searchTerm.trim())
            : items;
        if (renderItems.length) {
            component = (
                <FlatList
                    data={renderItems}
                    keyExtractor={item => item[uniqueKey]}
                    renderItem={rowData => this.getRow(rowData.item)}
                />
            );
        }
        return component;
    };

    render() {
        const { selector, textColor } = this.state;
        const { placeHolderText } = this.props;
        let listStyle = selector ? styles.selectorView : styles.dropdownView;
        let text;
        let items;

        if (selector) {
            text = <TextInput
                onChangeText={this.onChangeInput}
                placeholder={placeHolderText}
                placeholderTextColor={'#f0f1f1'}
                underlineColorAndroid="transparent"
                style={styles.container}
            />;
        } else {
            text = <Text
                style={[styles.container, {color: textColor}]}
                >{this.getSelectLabel()}
                </Text>;
        }

        if (selector) {
            items = <View>{this.renderItems()}</View>;
        }

        return (
            <View style={listStyle}>
                {<TouchableWithoutFeedback onPress={this.toggleSelector}>
                    <View style={styles.inputGroup}>
                        {text}
                        <View style={styles.dropdownIcon}>
                            <Image
                                source={selector ? usernameImg : passwordImg}
                                style={styles.indicator}
                            />
                        </View>
                    </View>
                </TouchableWithoutFeedback>}
                {items}
            </View>
        );
    }
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
    },
    indicator: {
        height: 10,
        width: 20,
    },
    dropdownIcon: {
        paddingTop: 10,
        paddingLeft: 10,
        paddingRight: 10,
        paddingBottom: 10,
    },
    selectedItem: {
        flexDirection: "row",
        alignItems: "center",
        paddingLeft: 15,
        margin: 3,
        borderRadius: 20,
        borderWidth: 2,
        justifyContent: "center",
        height: 40,
    },
    selectedItemText: {
        flex: 1,
        fontSize: 15,
    },
    selectorView: {
        flexDirection: "column",
        marginBottom: 10,
        elevation: 2,
    },
    inputGroup: {
        flexDirection: "row",
        alignItems: "center",
        paddingLeft: 16,
        backgroundColor: 'white',
        height: 40,
        borderBottomWidth: 1,
        borderBottomColor: 'black',
    },
    dropdownView: {
        flexDirection: "column",
        alignItems: "center",
        height: 40,
        marginBottom: 10,
    },
    menuListBox: {
        flexDirection: "row",
        flexWrap: "wrap",
    },
});

export default SelectDropDownList;

