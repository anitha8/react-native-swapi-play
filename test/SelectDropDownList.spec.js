import { Text, TouchableWithoutFeedback } from "react-native";
import React from "react";
import { configure, shallow } from "enzyme";

import Adapter from "enzyme-adapter-react-16";
import SelectDropDownList from "../src/components/SelectDropDownList";

configure({ adapter: new Adapter() });

describe("Testing SelectDropDownList Component", () => {
    let wrapper;
    const sampleItems = [{
        id: "92iijs7yta",
        name: "Ondo",
    }, {
        id: "a0s0a8ssbsd",
        name: "Ogun",
    }, {
        id: "16hbajsabsd",
        name: "Calabar",
    }, {
        id: "nahs75a5sg",
        name: "Lagos",
    }, {
        id: "667atsas",
        name: "Maiduguri",
    }, {
        id: "hsyasajs",
        name: "Anambra",
    }, {
        id: "djsjudksjd",
        name: "Benue",
    }, {
        id: "sdhyaysdj",
        name: "Kaduna",
    }, {
        id: "suudydjsjd",
        name: "Abuja",
    }];

    const props = {
        items: sampleItems,
        onSelectedItemsChange: jest.fn(),
        selectedItems: sampleItems.djsjudksjd,
        placeHolderText: "Type network name",
    };

    beforeEach(() => {
        wrapper = shallow(<SelectDropDownList {...props} />);
    });

    it("Check SelectDropDownList Snapshot", () => {
        expect(wrapper.dive()).toMatchSnapshot();
    });

    it("should update list expand and shrink flag", () => {
        const render = wrapper.dive();
        render.find(TouchableWithoutFeedback).simulate("press");
        expect(wrapper.state().selector).toBe(true);
        render.find(TouchableWithoutFeedback).simulate("press");
        expect(wrapper.state().selector).toBe(false);
    });

    it("should call renderItems when menu selected", () => {
        const render = wrapper.dive();
        const instance = wrapper.instance();
        const spy = jest.spyOn(instance, "renderItems");
        render.find(TouchableWithoutFeedback).simulate("press");
        expect(spy).toHaveBeenCalled();
        spy.mockReset();
        spy.mockRestore();
    });

    it("should update searchterm state when menu searched", () => {
        const render = wrapper.dive();
        render.find(TouchableWithoutFeedback).simulate("press");
        wrapper.instance().onChangeInput("Ca");
        expect(wrapper.state().searchTerm).toBe("Ca");
    });

    it("searchTerm reset when menu selected", () => {
        const render = wrapper.dive();
        render.find(TouchableWithoutFeedback).simulate("press");
        expect(wrapper.state().selector).toBe(true);
        wrapper.instance().toggleItem("Calabar");
        expect(wrapper.state().searchTerm).toBe("");
    });

    it("should call onSelectedItemsChange on item change", () => {
        const render = wrapper.dive();
        render.find(TouchableWithoutFeedback).simulate("press");
        expect(wrapper.state().selector).toBe(true);
        wrapper.instance().toggleItem("Calabar");
        expect(props.onSelectedItemsChange).toBeCalled();
    });

    it("check placeHolderText same as props", () => {
        const render = wrapper.dive();
        const text = render.find(Text).props().children;
        expect(text).toBe(props.placeHolderText);
    });
});
