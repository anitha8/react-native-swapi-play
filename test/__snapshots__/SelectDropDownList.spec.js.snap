// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing SelectDropDownList Component Check SelectDropDownList Snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "column",
        "height": 40,
        "marginBottom": 10,
      }
    }
  >
    <TouchableWithoutFeedback
      onPress={[Function]}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "white",
            "borderBottomColor": "black",
            "borderBottomWidth": 1,
            "flexDirection": "row",
            "height": 40,
            "paddingLeft": 16,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "flex": 1,
              },
              Object {
                "color": "#f0f1f1",
              },
            ]
          }
        >
          Type network name
        </Text>
        <View
          style={
            Object {
              "paddingBottom": 10,
              "paddingLeft": 10,
              "paddingRight": 10,
              "paddingTop": 10,
            }
          }
        >
          <Image
            source={
              Object {
                "testUri": "../../../src/images/password.png",
              }
            }
            style={
              Object {
                "height": 10,
                "width": 20,
              }
            }
          />
        </View>
      </View>
    </TouchableWithoutFeedback>
  </View>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <TouchableWithoutFeedback
          onPress={[Function]}
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "white",
                "borderBottomColor": "black",
                "borderBottomWidth": 1,
                "flexDirection": "row",
                "height": 40,
                "paddingLeft": 16,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Array [
                  Object {
                    "flex": 1,
                  },
                  Object {
                    "color": "#f0f1f1",
                  },
                ]
              }
            >
              Type network name
            </Text>
            <View
              style={
                Object {
                  "paddingBottom": 10,
                  "paddingLeft": 10,
                  "paddingRight": 10,
                  "paddingTop": 10,
                }
              }
            >
              <Image
                source={
                  Object {
                    "testUri": "../../../src/images/password.png",
                  }
                }
                style={
                  Object {
                    "height": 10,
                    "width": 20,
                  }
                }
              />
            </View>
          </View>
        </TouchableWithoutFeedback>,
        undefined,
      ],
      "style": Object {
        "alignItems": "center",
        "flexDirection": "column",
        "height": 40,
        "marginBottom": 10,
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "white",
                "borderBottomColor": "black",
                "borderBottomWidth": 1,
                "flexDirection": "row",
                "height": 40,
                "paddingLeft": 16,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Array [
                  Object {
                    "flex": 1,
                  },
                  Object {
                    "color": "#f0f1f1",
                  },
                ]
              }
            >
              Type network name
            </Text>
            <View
              style={
                Object {
                  "paddingBottom": 10,
                  "paddingLeft": 10,
                  "paddingRight": 10,
                  "paddingTop": 10,
                }
              }
            >
              <Image
                source={
                  Object {
                    "testUri": "../../../src/images/password.png",
                  }
                }
                style={
                  Object {
                    "height": 10,
                    "width": 20,
                  }
                }
              />
            </View>
          </View>,
          "onPress": [Function],
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Array [
                    Object {
                      "flex": 1,
                    },
                    Object {
                      "color": "#f0f1f1",
                    },
                  ]
                }
              >
                Type network name
              </Text>,
              <View
                style={
                  Object {
                    "paddingBottom": 10,
                    "paddingLeft": 10,
                    "paddingRight": 10,
                    "paddingTop": 10,
                  }
                }
              >
                <Image
                  source={
                    Object {
                      "testUri": "../../../src/images/password.png",
                    }
                  }
                  style={
                    Object {
                      "height": 10,
                      "width": 20,
                    }
                  }
                />
              </View>,
            ],
            "style": Object {
              "alignItems": "center",
              "backgroundColor": "white",
              "borderBottomColor": "black",
              "borderBottomWidth": 1,
              "flexDirection": "row",
              "height": 40,
              "paddingLeft": 16,
            },
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "accessible": true,
                "allowFontScaling": true,
                "children": "Type network name",
                "ellipsizeMode": "tail",
                "style": Array [
                  Object {
                    "flex": 1,
                  },
                  Object {
                    "color": "#f0f1f1",
                  },
                ],
              },
              "ref": null,
              "rendered": "Type network name",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": <Image
                  source={
                    Object {
                      "testUri": "../../../src/images/password.png",
                    }
                  }
                  style={
                    Object {
                      "height": 10,
                      "width": 20,
                    }
                  }
                />,
                "style": Object {
                  "paddingBottom": 10,
                  "paddingLeft": 10,
                  "paddingRight": 10,
                  "paddingTop": 10,
                },
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "source": Object {
                    "testUri": "../../../src/images/password.png",
                  },
                  "style": Object {
                    "height": 10,
                    "width": 20,
                  },
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        "type": [Function],
      },
      undefined,
    ],
    "type": "View",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <TouchableWithoutFeedback
            onPress={[Function]}
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "white",
                  "borderBottomColor": "black",
                  "borderBottomWidth": 1,
                  "flexDirection": "row",
                  "height": 40,
                  "paddingLeft": 16,
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Array [
                    Object {
                      "flex": 1,
                    },
                    Object {
                      "color": "#f0f1f1",
                    },
                  ]
                }
              >
                Type network name
              </Text>
              <View
                style={
                  Object {
                    "paddingBottom": 10,
                    "paddingLeft": 10,
                    "paddingRight": 10,
                    "paddingTop": 10,
                  }
                }
              >
                <Image
                  source={
                    Object {
                      "testUri": "../../../src/images/password.png",
                    }
                  }
                  style={
                    Object {
                      "height": 10,
                      "width": 20,
                    }
                  }
                />
              </View>
            </View>
          </TouchableWithoutFeedback>,
          undefined,
        ],
        "style": Object {
          "alignItems": "center",
          "flexDirection": "column",
          "height": 40,
          "marginBottom": 10,
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "white",
                  "borderBottomColor": "black",
                  "borderBottomWidth": 1,
                  "flexDirection": "row",
                  "height": 40,
                  "paddingLeft": 16,
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Array [
                    Object {
                      "flex": 1,
                    },
                    Object {
                      "color": "#f0f1f1",
                    },
                  ]
                }
              >
                Type network name
              </Text>
              <View
                style={
                  Object {
                    "paddingBottom": 10,
                    "paddingLeft": 10,
                    "paddingRight": 10,
                    "paddingTop": 10,
                  }
                }
              >
                <Image
                  source={
                    Object {
                      "testUri": "../../../src/images/password.png",
                    }
                  }
                  style={
                    Object {
                      "height": 10,
                      "width": 20,
                    }
                  }
                />
              </View>
            </View>,
            "onPress": [Function],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": Array [
                <Text
                  accessible={true}
                  allowFontScaling={true}
                  ellipsizeMode="tail"
                  style={
                    Array [
                      Object {
                        "flex": 1,
                      },
                      Object {
                        "color": "#f0f1f1",
                      },
                    ]
                  }
                >
                  Type network name
                </Text>,
                <View
                  style={
                    Object {
                      "paddingBottom": 10,
                      "paddingLeft": 10,
                      "paddingRight": 10,
                      "paddingTop": 10,
                    }
                  }
                >
                  <Image
                    source={
                      Object {
                        "testUri": "../../../src/images/password.png",
                      }
                    }
                    style={
                      Object {
                        "height": 10,
                        "width": 20,
                      }
                    }
                  />
                </View>,
              ],
              "style": Object {
                "alignItems": "center",
                "backgroundColor": "white",
                "borderBottomColor": "black",
                "borderBottomWidth": 1,
                "flexDirection": "row",
                "height": 40,
                "paddingLeft": 16,
              },
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "accessible": true,
                  "allowFontScaling": true,
                  "children": "Type network name",
                  "ellipsizeMode": "tail",
                  "style": Array [
                    Object {
                      "flex": 1,
                    },
                    Object {
                      "color": "#f0f1f1",
                    },
                  ],
                },
                "ref": null,
                "rendered": "Type network name",
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": <Image
                    source={
                      Object {
                        "testUri": "../../../src/images/password.png",
                      }
                    }
                    style={
                      Object {
                        "height": 10,
                        "width": 20,
                      }
                    }
                  />,
                  "style": Object {
                    "paddingBottom": 10,
                    "paddingLeft": 10,
                    "paddingRight": 10,
                    "paddingTop": 10,
                  },
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "source": Object {
                      "testUri": "../../../src/images/password.png",
                    },
                    "style": Object {
                      "height": 10,
                      "width": 20,
                    },
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                "type": [Function],
              },
            ],
            "type": [Function],
          },
          "type": [Function],
        },
        undefined,
      ],
      "type": "View",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
